### PR TITLE
Types/incidencematrix

### PR DESCRIPTION
**Pull request by alexej-jordan**
´pm_IncidenceMatrix´ and corresponding tests working.
had to adjust ´pm_Integer´ and ´pm_Rational´ for compatibility with Bool, just like its handled on julia side.

### CODE REVIEW

**Comment by kalmarek** [comment]
I will have a detailed look at this in Leipzig.

@benlorenz could You look at the C++ part?
@saschatimme could you look at julia part?

@alexej-jordan do I remember correctly that we tried to cast column/row to SparseVectors and it didn't work?

File: src/broadcast.jl [hunk 1]
```diff
@@ -2,18 +2,27 @@
     return pm_Matrix{T}(length(rows), length(cols))
 end
 
-Base.similar(X::Union{pm_Vector, pm_Matrix}, ::Type{S}, dims::Dims{1}) where
+@inline function pm_IncidenceMatrix{pm_NonSymmetric}(rows::UR, cols::UR) where {UR<:Base.AbstractUnitRange}
```
Comment (saschatimme) [inline src/broadcast.jl]: Why the `@inline`?

File: src/broadcast.jl [hunk 2]
```diff
@@ -2,18 +2,27 @@
...
+Base.similar(X::Union{pm_Vector, pm_Matrix, pm_IncidenceMatrix}, ::Type{S}, dims::Dims{1}) where {S} = Vector{S}(dims...)
 
-Base.similar(X::Union{pm_Vector, pm_Matrix}, ::Type{S}, dims::Dims{2}) where
+Base.similar(X::Union{pm_Vector, pm_Matrix, pm_IncidenceMatrix}, ::Type{S}, dims::Dims{2}) where
     {S <: pm_VecOrMat_eltypes} = pm_Matrix{convert_to_pm_type(S)}(dims...)
 
-Base.similar(X::Union{pm_Vector, pm_Matrix}, ::Type{S}, dims::Dims{2}) where {S} = Matrix{S}(dims...)
+Base.similar(X::Union{pm_Vector, pm_Matrix, pm_IncidenceMatrix}, ::Type{S}, dims::Dims{2}) where {S} = Matrix{S}(dims...)
+
+Base.similar(X::pm_IncidenceMatrix, ::Type{S}, dims::Dims{1}) where {S <: Bool} = BitArray{1}(undef, dims...)
```
Comment (saschatimme) [inline src/broadcast.jl]: `Bool` is a concrete type, so 

```
Base.similar(X::pm_IncidenceMatrix, ::Type{Bool}, dims::Dims{1}) = BitArray{1}(undef, dims...)
```
is more idiomatic in my opinion

File: src/incidencematrix.jl [hunk 3]
```diff
@@ -0,0 +1,81 @@
+using SparseArrays
+
+@inline function pm_IncidenceMatrix{pm_NonSymmetric}(mat::AbstractMatrix)
```
Comment (saschatimme) [inline src/incidencematrix.jl]: No `@inline` here and for the others

File: src/incidencematrix.jl [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+Base.size(m::pm_IncidenceMatrix) = (rows(m), cols(m))
+
+Base.@propagate_inbounds function Base.getindex(M::pm_IncidenceMatrix , i::Integer, j::Integer)
+    @boundscheck 1 <= i <= rows(M) || throw(BoundsError(M, [i,j]))
+    @boundscheck 1 <= j <= cols(M) || throw(BoundsError(M, [i,j]))
+    return _getindex(M, convert(Int64, i), convert(Int64, j))
+end
+
+Base.@propagate_inbounds function Base.setindex!(M::pm_IncidenceMatrix, val, i::Integer, j::Integer)
+    @boundscheck 1 <= i <= rows(M) || throw(BoundsError(M, [i,j]))
```
Comment (saschatimme) [inline src/incidencematrix.jl]: There is a handy base function
```suggestion
 @boundscheck checkbounds(M, i, j)
```

File: src/rationals.jl [hunk 5]
```diff
@@ -5,12 +5,12 @@ pm_Rational(a::BigInt, b::BigInt) = pm_Rational(pm_Integer(a), pm_Integer(b))
 pm_Rational(x::Rational{<:Integer}) = pm_Rational(numerator(x), denominator(x))
 
 pm_Rational(int::Integer) = pm_Rational(int, one(int))
-pm_Rational(x::Number) = pm_Rational(Rational(x))
+pm_Rational(x::Union{Number, Bool}) = pm_Rational(Rational(x))
```
Comment (saschatimme) [inline src/rationals.jl]: `Bool` is a subtype of `Number`, so I don't think this is needed.

File: src/rationals.jl [hunk 6]
```diff
@@ -23,7 +23,7 @@ for op in (:(==), :(<), :(<=))
 end
 
 function Base.promote_rule(::Type{<:pm_Rational},
-    ::Type{<:Union{Integer, Rational{<:Integer}}})
+    ::Type{<:Union{Integer, Rational{<:Integer}, Bool}})
```
Comment (saschatimme) [inline src/rationals.jl]:
```
julia> Bool <: Integer
true
```
So no `Bool` needed

File: src/rationals.jl [hunk 7]
```diff
@@ -59,4 +59,5 @@ Base.://(x::pm_Integer, y::Rational) = pm_Rational(x*numerator(y), denominator(y
 Base.://(x::pm_Rational, y::Union{Int8, Int16, BigInt, Unsigned}) = x//pm_Integer(y)
 Base.://(x::Union{Int8, Int16, BigInt, Unsigned}, y::pm_Rational) = pm_Integer(x)//y
 Base.://(x::pm_Rational, y::Rational{<:Integer}) = x//pm_Rational(y)
-Base.://(x::Rational{<:Integer}, y::pm_Rational) = pm_Rational(x)//y
+Base.://(x::Rational{<:Union{Integer, Bool}}, y::pm_Rational) = pm_Rational(x)//y
```
Comment (saschatimme) [inline src/rationals.jl]:
```suggestion
Base.://(x::Rational{<:Integer}, y::pm_Rational) = pm_Rational(x)//y
```

File: src/broadcast.jl [hunk 8]
```diff
@@ -24,3 +33,13 @@ function Base.similar(bc::Broadcast.Broadcasted{Broadcast.ArrayStyle{pm_Matrix}}
     ::Type{ElType}) where ElType
     return pm_Matrix{promote_to_pm_type(pm_Vector, ElType)}(axes(bc)...)
 end
+
+function Base.similar(bc::Broadcast.Broadcasted{Broadcast.ArrayStyle{T}},
+    ::Type{ElType}) where {T<:pm_IncidenceMatrix, ElType <: Bool}
```
Comment (kalmarek) [inline src/broadcast.jl]: Eltype is Bool

**Pull request merged**